### PR TITLE
Integration of Messaging feature

### LINKED ISSUE #38: IN/FE: Messaging
- Create a chat UI.
- A left sidebar that shows conversations with users, a chat UI to display messages and allow sending messages on the right
- Integrate with backend routes to send/receive messages

### PR DESCRIPTION
**Pull request by rj13371**
### What this PR does (required):
Integrates the frontend messaging component with the backend and allows users to send, receive, and start conversations with other users. 

### Screenshots / Videos (front-end only):
demo screenshot
demo screenshot

### Any information needed to test this feature (required):
Login with any user
Go to the messages tab
Currently you need to have the userId of another user to create a conversation but I assume we will handle starting conversations a different way with another ticket. It was not clear to me according to the mocks. 
click the start new conversation modal button and send a message to any user. 

The state will rerender when you send a new message and you will be able to immediately see it. 

Once you start a new conversation it will remain in your inbox and you will be able to send messages in the chat window. 

### Any issues with the current functionality (optional):
Was still having trouble with MUI, could not change the flex direction of the messages to display based on who sent the message. 
I could not get the scroll to start at bottom when you post a new message, but when you switch between tabs it saves the scroll position ref. 
The state will rerender when you send a new message and you will be able to immediately see it.

### CODE REVIEW

File: client/src/pages/Messages/MessagesDashboard.tsx [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+  }, [chatbox]);
+
+  const handleSubmit = (description: string, conversationId: string) => {
+    setSubmitting(true);
+    sendMessage(description, conversationId).then((data) => {
+      if (data.error) {
+        setSubmitting(false);
+        updateSnackBarMessage(data.error.message);
+      } else if (data.success) {
+        console.log(data.success);
```
Comment (sundayezeilo) [inline client/src/pages/Messages/MessagesDashboard.tsx:63]: Please let's get rid of all unnecessary logs to the console

**Review by sundayezeilo — COMMENTED** [review]
Hey @rj13371, great work with this one!

File: client/src/pages/Messages/CreateConversation.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      {isSubmitting ? (
+        <CircularProgress />
+      ) : (
+        <Stack spacing={3}>
+          <TextField sx={{ minWidth: '200px' }} id="message" label="message" value={message} onChange={handleChange} />
+
+          <TextField sx={{ minWidth: '50px' }} id="receiver" label="user" value={receiver} onChange={handleChange} />
+
+          <Button
+            onClick={() => handleSubmit(receiver, message)}
```
Comment (rajivtitus) [inline client/src/pages/Messages/CreateConversation.tsx:58]:
```suggestion
      type="submit"
```
Nit: The function will be invoked on form submission anyways, we don't have to pass an `onClick` prop

File: client/src/pages/Messages/Message/Message.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+interface Props {
+  message: MessageInterface;
+}
+
+export default function Message({ message }: Props): ReactElement {
+  const { loggedInUser } = useAuth();
+
+  const sender = message.sender.name === loggedInUser?.name ? true : false;
+
+  const classes = useStyles(sender);
```
Comment (rajivtitus) [inline client/src/pages/Messages/Message/Message.tsx:17]: 👍🏽

File: client/src/pages/Messages/Message/Message.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import { useAuth } from '../../../context/useAuthContext';
+
+interface Props {
+  message: MessageInterface;
+}
+
+export default function Message({ message }: Props): ReactElement {
+  const { loggedInUser } = useAuth();
+
+  const sender = message.sender.name === loggedInUser?.name ? true : false;
```
Comment (rajivtitus) [inline client/src/pages/Messages/Message/Message.tsx:15]: Maybe we should compare the logged in user's id with the participants. Right now we are comparing names, what if the logged in user changes his name from the Edit Profile form?

File: client/src/pages/Messages/MessagesDashboard.tsx [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        updateSnackBarMessage('Sent!');
```
Comment (rajivtitus) [inline client/src/pages/Messages/MessagesDashboard.tsx:64]: I don't think we would want to update the snackbar for every new message 😄

File: client/src/pages/Messages/MessagesDashboard.tsx [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+        getConversations().then((data) => {
```
Comment (rajivtitus) [inline client/src/pages/Messages/MessagesDashboard.tsx:65]: I don't think we need to be refetching all the conversations each time we send a message

File: client/src/pages/Messages/MessagesDashboard.tsx [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+  const [conversations, setConversations] = useState<Conversations>([]);
+  const [isSubmitting, setSubmitting] = useState(false);
+  const { updateSnackBarMessage } = useSnackBar();
+  const [chatbox, setChatBox] = useState<number>(0);
+  const { loggedInUser } = useAuth();
+  const [message, setMessage] = useState<string>('');
+  const scrollRef = useRef<HTMLLIElement>(null);
+  const [open, setOpen] = useState(false);
+  const handleOpen = () => setOpen(true);
```
Comment (rajivtitus) [inline client/src/pages/Messages/MessagesDashboard.tsx:43]:
```suggestion
  const handleOpen = () => setOpen((prevState) => !prevState);
```
State updates may be asynchronous: React sometimes batches multiple setState() calls for performance reasons. Let's use use a function which takes the previous state as a first argument.

File: client/src/pages/Messages/MessagesDashboard.tsx [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+        </Grid>
+        <Grid container component={Paper} className={classes.inboxSection}>
+          <Grid item xs={3} className={classes.borderRight500}>
+            <Divider />
+            <List>
+              {conversations.length &&
+                conversations.map((conversation, index) => (
+                  <ListItem button onClick={() => setChatBox(index)} key={conversation._id}>
+                    <ListItemIcon>
+                      <Avatar
```
Comment (rajivtitus) [inline client/src/pages/Messages/MessagesDashboard.tsx:130]: To increase reusability and code maintainability we could probably componentize the sender and receiver chat bubbles

**Comment by rajivtitus** [comment]
@rj13371 Very good work with this! Just some comments for you to have a look 👍🏽

**Pull request merged**